lib/src/cpp/utils/XMLManager.cpp

```diff
@@ -186,209 +184,353 @@ bool XMLManager::save_xml()
-                                real_index = std::stoi(*index);
-                            }
-                            catch (std::exception)
-                            {
-                                throw BadParameter(in + " could not be used as integer index");
-                            }
-
-                            // Transform index to real index
-                            if (real_index < 0)
-                            {
-                                real_index = index_list->size() + real_index;
-                            }
-
-                            // Check bounds
-                            if (real_index < 0 || real_index >= index_list->size())
-                            {
-                                // Throw ElementNotFound exception
-                                throw ElementNotFound(
-                                          tag_name + " does not have an element in position " +
-                                          std::to_string(real_index) + "\n");
-                            }
-                            // Return Node
-                            return node_list->item(index_list->at(real_index));
-                        }
-                    }
-                    // Complex Element
-                    else
-                    {
-                        found = true;
-                    }
-                }
+                // Parse index from string to int
+                real_index = std::stoi(index);
             }
-        }
-        // Throw exception if node not found
-        if (!found)
-        {
-            // Given element does not exist
-            throw ElementNotFound(
-                      tag_name + " does not have an element with key " + att_value + "\n");
+            catch (std::exception)
+            {
+                throw BadParameter(index + " could not be used as integer index");
+            }
+
+            // Transform index to real index
+            if (real_index < 0)
+            {
+                real_index = index_list->size() + real_index;
+            }
+
+            // Check bounds
+            if (real_index >= 0 && real_index < index_list->size())
+            {
+                // Return Node
+                last_node = node_list->item(index_list->at(real_index));
+                return;
+            }
+
+            // Set up exception message to be thrown if node should not be created
+            exception_message = xercesc::XMLString::transcode(last_node->getNodeName());
+            exception_message += " does not have an element in position " + std::to_string(real_index) + "\n";
         }
         else
         {
-            // Return Node
-            return tag_node;
+            // Set up exception message to be thrown if node should not be created
+            exception_message = xercesc::XMLString::transcode(last_node->getNodeName());
+            exception_message += " collection is empty\n";
         }
```

Comment:
I am not really sure if we should avoid this check and output always the same exception: `does not have element in position`. An empty collection fulfills that whichever index I pass the list does not have an element in that position. This way, the code could be made simpler and we have less exception checks. What do you think?

Reply:
I agree that we should keep all code as simple as possible, and your suggestion is pretty reasonable. User would always have the opportunity to use the `query` command to check collection length in those cases.